types/underscore/index.d.ts

```diff
@@ -95,10 +95,31 @@ declare module _ {
         (element: T, key: string, object: V): TResult;
     }
 
-    type CollectionIterator<T, TResult, V> = V extends List<T> ? ListIterator<T, TResult, V>
+    type CollectionIterator<T, TResult, V> =
+        V extends List<T> ? ListIterator<T, TResult, V>
         : V extends Dictionary<T> ? ObjectIterator<T, TResult, V>
         : never;
 
+    type Iteratee<V, R, T extends TypeOfCollection<V> = TypeOfCollection<V>> =
+        undefined |
+        CollectionIterator<T, R, V> |
+        keyof T |
+        string |
+        Array<string | number> |
+        Partial<T>;
+
+    type IterateeResult<V, I, T extends TypeOfCollection<V> = TypeOfCollection<V>> =
+        I extends undefined ? T // default iteratee is _.identity
+        : I extends CollectionIterator<T, infer R, V> ? R
+        : I extends keyof T ? T[I]
+        : I extends string | Array<string | number> ? any
+        : I extends Partial<T> ? boolean
+        : never;
+
+    type KeyOrPropertyName<T> = keyof T | string | number;
```

Comment:
Is this distinguishable from `string | number`?

Reply:
From [this experience](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r449374141) I had thought it would be necessary to include `keyof T` to be able to end up with an inferred `K` that was a string constant instead of just `string`, but when I removed it that didn't end up being the case. I was able to remove this from `Iteratee` as well, so I'm not sure if there was something odd about what I was doing in the linked example or if I'm just crazy 😕 

After re-reading what I had tried during that part of our discussion, I did re-verify that removing the `extends Iteratee<V, any>` constraint from `I` in `map` causes `I` to be `string` instead of `'a'` when `'a'` is provided as iteratee to help reassure myself of my sanity, but in simpler attempts to reproduce that behavior TS always end up going with the constant type instead of just `string` so I'm not sure precisely what's special about that case. Oh well; thanks for prompting me to double-check this!